src/backend/utils/misc/queryenvironment.c

```diff
@@ -906,40 +906,37 @@ ENRDropTempTables(QueryEnvironment *queryEnv)
 {
 	ListCell   *lc = NULL;
 	ObjectAddress object;
+	ObjectAddresses *objects;
 
 	if (!queryEnv)
 		return;
 
-	object.classId = RelationRelationId;
-	object.objectSubId = 0;
+	objects = new_object_addresses();
 
 	/*
 	 * Loop through the registered ENRs to drop temp tables.
 	 */
-	while (lc || (lc = list_head(queryEnv->namedRelList)))
+	foreach(lc, queryEnv->namedRelList)
 	{
-		EphemeralNamedRelation enr;
-
-		enr = (EphemeralNamedRelation) lfirst(lc);
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
 
-		if (enr->md.enrtype != ENR_TSQL_TEMP) {
-			lc = lnext(queryEnv->namedRelList, lc);
-			if (!lc)
-				break;
-			else
-				continue;
-		}
+		if (enr->md.enrtype != ENR_TSQL_TEMP)
+			continue;
 
-		/*
-		 * performDeletion() will remove the table AND the ENR entry, so no
-		 * need to remove the entry afterwards.
-		 */
+		object.classId = RelationRelationId;
+		object.objectSubId = 0;
 		object.objectId = enr->md.reliddesc;
-		performDeletion(&object, DROP_CASCADE, PERFORM_DELETION_INTERNAL | PERFORM_DELETION_QUIETLY);
-
-		/* Reset lc so we can read the head of the list */
-		lc = NULL;
+		add_exact_object_address(&object, objects);
 	}
+
+	/*
+	 * performMultipleDeletions() will remove the table AND the ENR entry,
+	 * so no need to remove the entry afterwards. Sort addresses beforehand
+	 * so that we don't run into dependency issues when dropping objects.
+	 */
+	sort_object_addresses(objects);
```

Comment:
Did we try this with temp table with identity and primary key?
I am curious how this will work if index is also in enr.

when we do something like `create table #t (id int identity primary key)` then the flow is

create sequence
create table
create index

so after reverse sort, we try to drop index first. Can we quickly check this works?

Thanks

Reply:
Thanks for catching that, it actually does run into issues trying to drop the index first. 

```
1> create procedure pz as
2> begin
3> create table #t (id int identity primary key)
4> create index i on #t(id)
5> end
6> go
1> exec pz
2> go
Msg 3723, Level 16, State 1, Server BABELFISH, Line 1
cannot drop index "#t_pkey" because constraint #t_pkey on table "#t" requires it
```

Reply:
what is the final conclusion here?

Reply:
I would say that, the current change is preferable to the old behavior because using just sequence in TV is far more prevalent than this edge case of creating index on temp table with primary key.

Reply:
Is it possible to come up with a solution that covers both?

Reply:
Interestingly enough, removing the `sort_object_addresses` call seems to have fixed both issues. Seems like performMultipleDeletions correctly gets the order to delete objects in. See tests in https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/1708